vms/platformvm/block/executor/block.go

```diff
@@ -29,34 +29,49 @@ func (*Block) ShouldVerifyWithContext(context.Context) (bool, error) {
 	return true, nil
 }
 
-func (b *Block) VerifyWithContext(_ context.Context, ctx *smblock.Context) error {
-	pChainHeight := uint64(0)
-	if ctx != nil {
-		pChainHeight = ctx.PChainHeight
-	}
-
+func (b *Block) VerifyWithContext(ctx context.Context, blockContext *smblock.Context) error {
 	blkID := b.ID()
-	if blkState, ok := b.manager.blkIDToState[blkID]; ok {
-		if !blkState.verifiedHeights.Contains(pChainHeight) {
-			// PlatformVM blocks are currently valid regardless of the ProposerVM's
-			// PChainHeight. If this changes, those validity checks should be done prior
-			// to adding [pChainHeight] to [verifiedHeights].
-			blkState.verifiedHeights.Add(pChainHeight)
+	blkState, previouslyExecuted := b.manager.blkIDToState[blkID]
+	warpAlreadyVerified := previouslyExecuted && blkState.verifiedHeights.Contains(blockContext.PChainHeight)
+
+	// If the chain is bootstrapped and the warp messages haven't been verified,
+	// we must verify them.
+	if !warpAlreadyVerified && b.manager.txExecutorBackend.Bootstrapped.Get() {
+		err := VerifyWarpMessages(
+			ctx,
+			b.manager.ctx.NetworkID,
+			b.manager.ctx.ValidatorState,
+			blockContext.PChainHeight,
+			b,
+		)
+		if err != nil {
+			return err
 		}
+	}
 
-		// This block has already been verified.
+	// If the block was previously executed, we don't need to execute it again,
+	// we can just mark that the warp messages are valid at this height.
+	if previouslyExecuted {
+		blkState.verifiedHeights.Add(blockContext.PChainHeight)
```

Comment:
why we only mark this height as verified if it's previouslyExecuted? Do we expect VerifyWithContext to be always called multiple times? Or do we handle verifiedHeights later at same point?